### PR TITLE
MINIFI-217: Implement namespaces. 

### PR DESCRIPTION
**Pull request by phrocker**
Updates namespaces and removes use of raw pointers for user facing API.

Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [x] Does your PR title start with MINIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the LICENSE file?
- [x] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by phrocker** [comment]
Some suggestions I've found for myself. Maybe rename utils to util. 

I have not added a provenance repository. Will do that in a subsequent PR or here if desired.

**Comment by jdye64** [comment]
@phrocker when compiling I'm seeing a couple of warnings that look like

`
Serializable.cpp:190:21: warning: comparison of constant 2047 with expression of type 'char' is always false
        } else if (c > 2047){`

Any thoughts on this?

**Comment by phrocker** [comment]
@jdye64 Since I moved to using auto instead of int, we are actually using a car to iterate through the string, therefore the previous code that existed now throws the warning. This is a placeholder for a block I submitted. I think it is safe to treat this as UTF-8 since NiFi enforces UTF-8, but before doing so I'd like some more testing. Therefore I leave the warning be as a placeholder. Realistically we could just change auto to int to get rid of the warning, but I think it serves as a warning that we weren't handling UTF correctly previously and now. I will deal with this in a subsequent PR. 

Again, I think we're okay as long as NiFi sticks with UTF-8, but I'd like to beef up the tests.

File: libminifi/include/Connection.h [hunk 1]
```diff
@@ -28,174 +28,168 @@
...
+    return !isEmpty();
+  }
+
+  bool isRunning() {
+    return true;
+  }
 
+ protected:
+  //! Source Processor UUID
+  uuid_t _srcUUID;
```
Comment (jdye64) [inline libminifi/include/Connection.h]: Can you move "_" to end of class data member to adhere to Google C++ Style guid?

File: libminifi/include/Connection.h [hunk 2]
```diff
@@ -28,174 +28,168 @@
...
+  //! Source Processor (ProcessNode/Port)
+  std::shared_ptr<core::Connectable> _srcProcessor;
+  //! Destination Processor (ProcessNode/Port)
+  std::shared_ptr<core::Connectable> _destProcessor;
+  //! Max queue size to apply back pressure
+  std::atomic<uint64_t> _maxQueueSize;
+  //! Max queue data size to apply back pressure
+  std::atomic<uint64_t> _maxQueueDataSize;
+  //! Flow File Expiration Duration in= MilliSeconds
+  std::atomic<uint64_t> _expiredDuration;
```
Comment (jdye64) [inline libminifi/include/Connection.h]: Can you move "_" to end of class data member to adhere to Google C++ Style guid? Same for all variables here.

File: libminifi/include/FlowControlProtocol.h [hunk 3]
```diff
@@ -45,297 +51,269 @@ class FlowController;
...
+    configure_ = Configure::getConfigure();
+    _socket = 0;
+    _serverName = "localhost";
+    _serverPort = DEFAULT_NIFI_SERVER_PORT;
+    _registered = false;
+    _seqNumber = 0;
+    _reportBlob = NULL;
+    _reportBlobLen = 0;
+    _reportInterval = DEFAULT_REPORT_INTERVAL;
+    _running = false;
```
Comment (jdye64) [inline libminifi/include/FlowControlProtocol.h]: Can you move "_" to end of class data member to adhere to Google C++ Style guide? Same for all variables here.

File: libminifi/include/FlowFileRecord.h [hunk 4]
```diff
@@ -32,198 +32,138 @@
...
+   //! UUID string
+   std::string _uuidStr;
+   //! UUID string for all parents
+   std::set<std::string> _lineageIdentifiers;
+   */
+
+ private:
+
+  //! Local flow sequence ID
+  static std::atomic<uint64_t> _localFlowSeqNumber;
```
Comment (jdye64) [inline libminifi/include/FlowFileRecord.h]: Can you move "_" to end of class data member to adhere to Google C++ Style guide?

File: libminifi/include/SchedulingAgent.h [hunk 5]
```diff
@@ -28,72 +28,90 @@
...
+class SchedulingAgent {
+ public:
+  //! Constructor
+  /*!
+   * Create a new processor
+   */
+  SchedulingAgent(std::shared_ptr<provenance::ProvenanceRepository> repo) {
+    configure_ = Configure::getConfigure();
+    logger_ = logging::Logger::getLogger();
+    _running = false;
```
Comment (jdye64) [inline libminifi/include/SchedulingAgent.h]: Can you move "_" to end of class data member to adhere to Google C++ Style guide? Same for all variables here.

File: libminifi/include/SchedulingAgent.h [hunk 6]
```diff
@@ -28,72 +28,90 @@
...
+  //! Configure
+  Configure *configure_;
+  //! Mutex for protection
+  std::mutex _mtx;
+  //! Whether it is running
+  std::atomic<bool> _running;
+  //! AdministrativeYieldDuration
+  int64_t _administrativeYieldDuration;
+  //! BoredYieldDuration
+  int64_t _boredYieldDuration;
```
Comment (jdye64) [inline libminifi/include/SchedulingAgent.h]: Can you move "_" to end of class data member to adhere to Google C++ Style guide? Same for all variables here.

**Comment by brosander** [comment]
@phrocker 

From a Java dev perspective, the namespaces look a little shallower than the Java package names.  I think that's probably justifiable though to prevent a ton of boilerplate, given that the number of classes is much lower in cpp-land.

Big fan of namespacing to avoid name clashes and just generally keep different ideas separate from each other.  Think this looks like a great step forward :smile:

**Comment by phrocker** [comment]
Also added a fix for MINIFI-236 since that impeded my testing.

**Comment by phrocker** [comment]
Updated to facilitate changes for #67

**Comment by apiri** [comment]
starting review

**Comment by apiri** [comment]
hey @phrocker:

changes and structure seem reasonable to me
build works but seem to be having some troubles with running in current form under default config:

Received the following on OS X 10.11.6 
> libc++abi.dylib: terminating with uncaught exception of type YAML::TypedBadConversion<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >: bad conversion

and the on CentOS 7
>terminate called after throwing an instance of 'YAML::TypedBadConversion<std::string>'
  what():  bad conversion

**Comment by apiri** [comment]
Appears this actually cropped up sometime earlier.  Doing the bisect routine to figure it out.

**Comment by apiri** [comment]
Looks like it was introduced in c4fea0cd94f9ea2ae94303ff5b574c0c00d6745a.

**Comment by apiri** [comment]
Thanks for taking the time to adjust.  Looks like this rights the ship and should be good to go.  Will get merged.

**Pull request closed without merging**